features/facebookImport/src/model/storage.js

```diff
@@ -37,15 +37,33 @@ export class ZipFile {
     constructor(file, pod) {
         this._pod = pod;
         this._file = file;
+
+        this._entriesList = null;
+        this._entriesSet = null;
     }
 
     get id() {
         return this._file.id;
     }
 
-    async getEntries() {
+    async _readEntriesList() {
         const { polyOut } = this._pod;
-        return polyOut.readdir(this.id);
+        return await polyOut.readdir(this.id);
+    }
+
+    async refreshCachedEntries() {
```

Comment:
Is there actually any use case for invalidating the cache later on? All I see is that whenever we create a `ZipFile`, we need to fall that function.

I'd prefer it if `getEntries()` would actually handle the caching, i.e. something like:

```
async getEntries() {
    if (!this._entries) this._entries = this._readEntries();
    return this._entries;
}
```

If we do need to invalidate the cache, we could add a function called `invalidateEntriesCache` that sets `this._entries` back to `null`, but like I said, don't currently see if we'd need it.

Reply:
Right now we do not need cache invalidation here. `ZipFile` is created and used only while running importers or mini-stories and discarded afterwards. 

In an idea case the storage would hold a list of `ZipFile` entries and manage their lifecycle. But wanted to avoid doing larger refactorings.

In the current design the cache has to be set before calling `getEntries` or `hasEntries`. These methods are called in parallel by the importers and if we set the cache there we need to handle concurrency issues.  Might be simpler to just have `getEntries` as below and assume `ZipFile`  is initialised correctly.
 
```
 async getEntries() {
        return this._entriesList;
  }
```

Reply:
Well, if we really need to have _both_ an array and a set, we need a bit more ceremony, but we could still do it lazily to be far less error prone. Something like:

```
// TODO: Rename to getEntries, rename current getEntries to getEntriesArray
async getEntriesSet() {
    if (!this._entries) await this._entries = this._readEntries();
    return this._entries;
}

// Deprecated
async getEntries() {
    const entries = await this.getEntriesSet();
    return [...entries];
}

// Deprecated
async hasEntry(entry) {
    const entries = await this.getEntriesSet();
    return entries.has(entry);
}
```

That magic function one needs to call after creating a `ZipFile` just seems _too_ hacky to me :D That's just trouble asking to happen.

Reply:
I can't imagine how there'd be any concurrency going on, the browser engines we're running this code in execute it single threaded.

Reply:
But yeah, I'd prefer to get rid of `getEntries` returning an array - and the need for `hasEntry` while at it. One approach to do this could be to add some functions to the object returned by `getEntries()`, so that it implements the array functions we use elsewhere in the code, if you really don't want to refactor it now :D

Reply:
Somehow, I still think we will get errors dus to parallel execution. To run all importers we use `Promise.all` and in each importer we get the list of files. Wouldn't that end up calling the getEntries methods from multiple promises at the same time? Even if it doesn't we no longer control when the call happens.

One easy improvement is to add a static method for instantiating a zip file. The API for creating a zip file can be `ZipFile.createForData(...)` and that does all initialisation. After that we assume all data is there.

Reply:
But the more I think about it, I somehow end up thinking that the clients of a `ZipFile` should not care how the list of entries is stored internally. For example, now each entry is a string, but might be better to change it later on to a list of ZipEntry objects.

Why now add an api like:
- `getEntries`: returns something that can iterate over entries
- `hasEntry`: check that an entry exists
- `entriesCount`: returns the number of entires

With this  we can have a set, a hash map or an array to store the entities.

Reply:
The static method works for me!

Not a huge fan of creating a custom iterable-like API, JavaScript isn't really made for that :D `getEntries` would be our leaky abstraction here, and callers have to know that even if they _can_ get an array, they're not supposed to use it for checking whether an entry is there, or to see how many entries there are. It's weird.

A set is less of an implementation detail in my book, and more a semantic data structure that also tells the caller that entries are unique.

I'm OK with implementing some workaround now so we don't have to update all usages of `getEntries`, but I do think we should design it so that it's clear that this is a temporary workaround.

Reply:
Javascript is not parallel, it simply puts the events in order and returns promises, but everything is run in a single thread.

Reply:
That's true for code execution but not IO, if you kick off two network requests they do happen in parallel e.g. I believe the same is true for polyPod APIs, didn't see a mutex anywhere in the native code. But even if it's just consecutive, it'd be at least one unnecessary `readdir` call with my suggested approach.